### ISSUE TITLE
Throwing ListSizeMismatch instead of timeout in CollectionCondition

### ISSUE THREAD

**Issue by Stefan1972** [codeborne/selenide#309]
Sample Code:
```
ElementsCollection newNameCell=pluginsTable.$$(".qooxdoo-table-cell").filterBy(text("TestingProxy_1"));
newNameCell.shouldHaveSize(1);
```
If timeout is exeeded this error occurs (note ElementsCollection.java:105):
```
ListSizeMismatch : expected: 1, actual: 1, collection: <div>/.qooxdoo-table-cell.filter(text 'TestingProxy_1')
Elements: [
	<div class="qooxdoo-table-cell">TestingProxy_1</div>
]
Timeout: 6 s.
	at com.codeborne.selenide.collections.ListSize.fail(ListSize.java:24)
	at com.codeborne.selenide.ElementsCollection.waitUntil(ElementsCollection.java:105)
	at com.codeborne.selenide.ElementsCollection.should(ElementsCollection.java:56)
	at com.codeborne.selenide.ElementsCollection.shouldHave(ElementsCollection.java:49)
	at com.codeborne.selenide.ElementsCollection.shouldHaveSize(ElementsCollection.java:34)
	at ***.uitest.pageobjects.ScriptProxy.getPluginCell(ScriptProxy.java:60)
```
Html snippet:
```
<div style="font-family: Tahoma,"Liberation Sans",Arial; font-size: 11px; line-height: 1.4; background-color: rgb(240, 240, 240); color: rgb(26, 26, 26); border-bottom: 1px solid rgb(231, 231, 231); position: relative; height: 19px; width: 100%;">
<div class="qooxdoo-table-cell" style="left:0px;width:97px;height:20px;">
<div style="margin-top:2px;text-align:center;vertical-align:middle;background:rgb(22, 171, 57);color:white;">enabled</div>
</div>
<div class="qooxdoo-table-cell" style="left:110px;width:162px;height:20px;">TestingProxy_1</div>
<div class="qooxdoo-table-cell" style="left:285px;width:122px;height:20px;">Script Proxy</div>
<div class="qooxdoo-table-cell" style="left:420px;width:42px;height:20px;"></div>
<div class="qooxdoo-table-cell" style="left:475px;width:287px;height:20px;">Testing proxy.</div>
<div class="qooxdoo-table-cell" style="left:775px;width:137px;height:20px;">TestingProxy.js</div>
</div>
```

Expectation whould be that an timeout error occurs.

What i don't understand is that the timeout occurs after 6 seconds but i set the timeout to 10 seconds. Seems like it's getting the default timeout somehow ...

**Comment by asolntsev** [comment]
@Stefan1972 In Selenide 3.3 we introduced a separate timeout for collections. 
See http://selenide.org/2016/02/11/selenide-3.3

Default value is 6 seconds.

**Comment by Stefan1972** [comment]
Wow! That was a quick answer. Thanks! This helps a lot.

Anyway: In my opinion this is still an error: When an timeout occurs it should be named so.

**Comment by asolntsev** [comment]
@Stefan1972 In this case, all Selenide methods should throw TimeoutException. Because all Selenide methods wait for a few seconds until the expected condition becomes true. 

Actually the wording `ListSizeMismatch` is very exact: it says that list size is wrong, and "yes, I have been waiting for a while, it's still wrong".

**Comment by hujog** [comment]
The point is: The ListSizeMismatch is wrong. 1 Element is expected and 1 Element is found and still a ListSizeMismatch error is thrown ("ListSizeMismatch : expected: 1, actual: 1")


```
ListSizeMismatch : expected: 1, actual: 1, collection: <div>/.qooxdoo-table-cell.filter(text 'TestingProxy_1')
Elements: [
    <div class="qooxdoo-table-cell">TestingProxy_1</div>
]
```

**Comment by Stefan1972** [comment]
@hujog thanks. that's the point.

@asolntsev 
Please take look at ElementsCollection.waitUntil(CollectionCondition condition, long timeoutMs) (line 76-106):
```
  protected void waitUntil(CollectionCondition condition, long timeoutMs) {
    lastError = null;
    final long startTime = System.currentTimeMillis();
    boolean conditionMatched = false;
    do {
      try {
        actualElements = collection.getActualElements();
        if (condition.apply(actualElements)) {
          if (conditionMatched) {
            return;
          } else {
            conditionMatched = true;
            sleep(collectionsPollingInterval);
            continue;
          }
        } else {
          conditionMatched = false;
        }
      } catch (WebDriverException elementNotFound) {
        lastError = elementNotFound;

        if (Cleanup.of.isInvalidSelectorError(elementNotFound)) {
          throw Cleanup.of.wrap(elementNotFound);
        }
      }
      sleep(pollingInterval);
    }
    while (System.currentTimeMillis() - startTime < timeoutMs);

    condition.fail(collection, actualElements, lastError, timeoutMs);
  }
```
If the timeout exceeds during sleep(collectionsPollingInterval) and the continue statement but the Condition matches , then described behavior occurs.

To reproduce the behavior just set the collectionsPollingInterval sligtly bigger then the collectionsTimeout i.e.
collectionsTimeout = 3000;
collectionsPollingInterval = 3100;

Note: The behavior also occurs with "normal" setting. The probability is just much lower... To evaluate that you can set a conditional breakpoint at the continue statement (line 89):
`actualElements != null && actualElements.size() == 1 && System.currentTimeMillis() - startTime >= timeoutMs`

**Comment by asolntsev** [comment]
@Stefan1972 Wow! Thank you for such a detailed error description, and especially for "steps to reproduce". Now it's clear how this error happens, and it's easy to fix it.

**Comment by asolntsev** [comment]
@Stefan1972 Thank you! Fixed in Selenide 3.6

**Comment by Stefan1972** [comment]
Nice. Keep up the good work!
